---
title: 'scPipe_atac report `r params$organism` scPipe `r Biobase::package.version("scPipe")`'
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: no
params:
  input_folder: "FOLDER__"
  organism: "ORGANISM__"
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                       # results = "hide",
                       message = FALSE,
                       warning = FALSE)
```

```{r}
library(data.table)
library(dplyr)
library(ggplot2)
library(kableExtra)
library(RColorBrewer)
library(grid)
library(knitr)
library(stringr)
theme_set(theme_bw())

```

# Global quality statistics

```{r}
trimbarcode_stats_filename    <- file.path(params$input_folder, "scPipe_atac_stats", "stats_file_trimbarcode.txt")
trimbarcode_stats             <- read.csv(trimbarcode_stats_filename, header=FALSE) %>% as.data.frame()
trimbarcode_stats             <- stringr::str_replace_all(trimbarcode_stats[,1], "\\: +", "\\,") %>% as.data.frame()
trimbarcode_stats             <- stringr::str_split_fixed(trimbarcode_stats[,1], ",", n=2) %>% as.data.frame()

knitr::kable(trimbarcode_stats,
             col.names = c('Statistic', 'Value'),
             caption = "Global quality statistics") %>%
    kableExtra::kable_styling()
```

# Alignment statistics

```{r}
alignment_stats_filename <- file.path(params$input_folder, "scPipe_atac_stats", "stats_file_align.txt")
alignment_stats          <- read.csv(alignment_stats_filename, header = FALSE, skip = 1) %>% as.data.frame()
knitr::kable(alignment_stats, 
             col.names = c('Statistic', 'Value'), 
             caption = "Global alignment statistics") %>%
    kable_styling()
```

Below is a stacked bar plot showing the distribution of how the fragments have been mapped. The left plot shows the number of uniquely and multi-mapped fragments, while the right plot provides a more detailed breakdown of the mapped fragments.

```{r alignment-stats-plot, fig.wide = TRUE, fig.asp = 0.9}
as.1 <- alignment_stats[alignment_stats$V1 %in% c("Unmapped_fragments", 
                                                  "Uniquely_mapped_fragments", 
                                                  "Multi_mapping_fragments"),] %>% 
  mutate(Type = "1") %>% rename(Mapping_type = "V1")

as.2 <- alignment_stats[alignment_stats$V1 %in% c("Properly_paired_fragments", 
                                                  "Singleton_fragments",
                                                  "More_than_one_chr_fragments",
                                                  "Unexpected_strandness_fragments",
                                                  "Unexpected_template_length",
                                                  "Inversed_mapping"),] %>% 
  mutate(Type = "2") %>% rename(Mapping_type = "V1")

as.1$Proportion <- as.1$V2/sum(as.1$V2)
as.2$Proportion <- as.2$V2/sum(as.2$V2)

as.combined <- rbind(as.1, as.2)
g1 <- ggplot(as.1, aes(x = Type, y = Proportion, fill = Mapping_type)) + 
  geom_bar(stat="identity", width=0.25) +
  ylab("Proportion of total fragments") + 
  ggtitle("Uniquely mapped, multimapped and \nunmapped fragments") +
  theme(axis.text.x=element_blank(),
        title = element_text(size=10),
        axis.ticks.x=element_blank(),
        axis.title.x=element_blank()) +
  scale_y_continuous(labels = scales::percent_format(), expand = expansion(mult = c(0, 0)))

g2 <- ggplot(as.2, aes(x = Type, y = Proportion, fill = Mapping_type)) + 
  geom_bar(stat="identity", width=0.25) +
  ylab("Proportion of total fragments") + 
  ggtitle("Breakdown of mapped fragments") +
  theme(axis.text.x=element_blank(),
        title =element_text(size=10),
        axis.ticks.x=element_blank(),
        axis.title.x=element_blank()) +
  scale_y_continuous(labels = scales::percent_format(), expand = expansion(mult = c(0, 0))) 

gridExtra::grid.arrange(g1, g2, ncol=2)
```

## Alignment statistics per chromosome

```{r}
alignment_stats_chrom_filename <- file.path(params$input_folder, "scPipe_atac_stats", "stats_file_align_per_chrom.csv")
alignment_stats_chrom          <- read.csv(alignment_stats_chrom_filename)
alignment_stats_chrom <- alignment_stats_chrom[!(alignment_stats_chrom$seqnames %in% c("*")),  !(names(alignment_stats_chrom) %in% c("unmapped"))]
DT::datatable(alignment_stats_chrom)
```

### Read count distribution across genome

A plot showing the read count distribution across chromosomes is shown below.

```{r}
alignment_stats_chrom <- rename(read.csv(alignment_stats_chrom_filename), reads = "mapped") %>% 
  filter(reads > 0) 
alignment_stats_chrom <- alignment_stats_chrom[, !(names(alignment_stats_chrom) %in% c("unmapped"))]
# Plot of reads per barcode 
alignment_stats_chrom <- alignment_stats_chrom[!(alignment_stats_chrom$seqnames %in% c("chrM")), ] # Remove chrM
RPK <- alignment_stats_chrom$reads/alignment_stats_chrom$seqlength # Reads per kilo-base (normalize by chromosome length)
PMSF <- sum(RPK)/10^6 # Per million scaling factor (normalize by total RPK)
alignment_stats_chrom$TPM <- RPK/PMSF # Transcripts per million
alignment_stats_chrom <- alignment_stats_chrom %>% arrange(TPM)
# Calculate widths
alignment_stats_chrom$widths <- alignment_stats_chrom$seqlength/sum(alignment_stats_chrom$seqlength)
alignment_stats_chrom$locs <- cumsum(alignment_stats_chrom$widths) - alignment_stats_chrom$widths/2
ggplot(alignment_stats_chrom,
       aes(x=locs,
           y=TPM,
           width=widths)) +
  geom_bar(stat = "identity", fill = "#1B9E77", color = "#1B9E77") +
  scale_fill_brewer(palette="Set1") +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_y_continuous(limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.1))) +
  xlab("Chromosome") +
  ylab("Transcripts per million") +
  ggtitle("Read count (TPM) distribution across chromosomes")
```

# Barcode demultiplexing statistics

Shows the alignment statistics for reads with matched (without or with error correction) and unmatched barcodes.


```{r}
data <- read.csv(file.path(params$input_folder, "scPipe_atac_stats", "demultiplexing_stats.csv"))

DT::datatable(data, options=list(paging=FALSE, searching=FALSE))
```

```{r}

data$prop <- data$count/sum(data$count) 

ggplot(data, aes_string(x="status", y="prop", fill="status")) + 
  scale_fill_brewer(palette="Set1") +
  geom_bar(stat="identity", width=1) +
  geom_text(aes(y = prop, label = percent(prop)), vjust = -0.5) +
  scale_y_continuous(expand = expansion(mult = c(0, 0))) +
  theme(axis.text.x=element_text(angle = 50, hjust = 1),
        axis.ticks.x=element_blank(),
        panel.border = element_blank()) +  
  xlab("status") +
  ylab("percentage") +
  expand_limits(y = 1) +
  ggtitle(paste0("Overall alignment mapping statistics of demultiplexed reads")) 
```

A stacked bar plot showing the alignment status of reads in a random sample of 500 barcodes (including error-corrected ones) is given below:

```{r barcode-alignment-dist, fig.wide = TRUE, fig.asp = 0.9}
stats_file <- file.path(params$input_folder, "scPipe_atac_stats", "mapping_stats_per_barcode.csv")
data <- read.csv(stats_file, row.names = "barcode")[-5] %>% sample_n(500) 
mapping_stat <- data %>% arrange(mapped)
mapping_stat$barcode <- stats::reorder(rownames(mapping_stat), mapping_stat$mapped)
mapping_stat_prop <- as.data.frame(prop.table(as.matrix(mapping_stat[, sapply(mapping_stat, is.numeric)]), 1))
mapping_stat_prop$barcode <- mapping_stat$barcode
dat.m <- melt(mapping_stat, id.vars="barcode")
dat.m1 <- melt(mapping_stat_prop, id.vars="barcode")
colnames(dat.m)[2] <- "type"
colnames(dat.m1)[2] <- "type"
ggplot(dat.m1, aes_string(x="barcode", y="value", fill="type")) + scale_fill_brewer(palette="Set1")+
  geom_bar(stat="identity", width=1)+
  xlab("barcode") +
  ylab("percentage of reads")+
  scale_y_continuous(labels = percent_format(), expand = expansion(mult = c(0, 0))) +
  
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

# Duplicate removal statistics 

A table showing the duplicate removal statistics (requires samtools 1.10 or greater).
```{r}

duplicate_stats.file <- file.path(params$input_folder, "scPipe_atac_stats", "duplicate_removal_stats.txt")
if (file.exists(duplicate_stats.file)) {
  duplicate.stats <- read.table(duplicate_stats.file, sep=":", skip=1)
  DT::datatable(duplicate.stats)
} else {
  cat("duplicate_removals_stats.txt couldn't be located in the stats folder. Did you run the duplicate removal function with samtools >= 1.10?\n")
}

```

# Per cell statistics

```{r}
sce <- readRDS(file.path(params$input_folder, "scPipe_atac_SCEobject.rds"))
```

<!-- Figures \@ref(fig:percell-counts) and \@ref(fig:percell-features) -->

## Counts per cell

A histogram of the total number of fragments across all called cells.
```{r percell-counts, fig.wide = TRUE, fig.asp = 0.9}
sc_atac_plot_fragments_per_cell(sce)

```

## Features per cell

A histogram of the number of overlapping features across all called cells.
```{r percell-features, fig.wide = TRUE, fig.asp = 0.9}
sc_atac_plot_features_per_cell(sce)

```

Plot showing the number of overlapping features across all cells, ordered.
```{r}
sc_atac_plot_features_per_cell_ordered(sce)

```

# Per feature statistics

## Counts per feature

A histogram of the number of overlapping fragments across all features.
```{r perfeature-counts, fig.wide = TRUE, fig.asp = 0.9}
sc_atac_plot_fragments_per_feature(sce)

```

## Cells per feature

A histogram of the number of associated cells for each feature.
```{r perfeature-cells, fig.wide = TRUE, fig.asp = 0.9}
sc_atac_plot_cells_per_feature(sce)

```


## Counts vs. features per cell

A scatter plot showing the relationship between the number of fragments and number of overlapping features for all selected cells.
```{r corr-cells, fig.wide = TRUE, fig.asp = 0.9}
sc_atac_plot_fragments_features_per_cell(sce)

```

## Counts vs. cells per feature

A scatter plot showing the relationship between the number of overlapping fragments and associated cells for each feature.
```{r corr-features, fig.wide = TRUE, fig.asp = 0.9}
sc_atac_plot_fragments_cells_per_feature(sce)

```



# Quality control plots

```{r}

anno.folder <- system.file("extdata/annotations", package = "scPipe", mustWork = TRUE) 

frags_file <- file.path(params$input_folder, "fragments.bed")
peaks_file <- file.path(params$input_folder, "NA_peaks.narrowPeak")
promoters_file <- file.path(anno.folder, "hg38_promoter.bed")
tss_file <- file.path(anno.folder, "hg38_tss.bed")
enhs_file <- file.path(anno.folder, "hg38_enhancer.bed")
qc_per_bc_file <- file.path(params$input_folder, "qc_per_bc_file.txt")

unfiltered_mtx <- readRDS(file.path(params$input_folder, "unfiltered_feature_matrix.rds"))
unfiltered_mtx_bcs <- colnames(unfiltered_mtx)

frags <- fread(frags_file)[V4 %in% unfiltered_mtx_bcs, ]

# read in qc_per_bc file
bc_stat <- fread(qc_per_bc_file)[bc %in% unfiltered_mtx_bcs, ]

# read in called cellular barcodes
cell_barcode_file <- file.path(params$input_folder, "non_empty_barcodes.txt")
cell_barcodes <- fread(cell_barcode_file, header=F)$V1

qc_sele <- bc_stat[bc %in% cell_barcodes, ]
qc_nonsele <- bc_stat[!bc %in% cell_barcodes, ]
```

## Total no. of fragments vs. fraction in peaks

A scatter plot of the total number of fragments and the fraction of them overlapping with peaks for cells and non-cells.
```{r  frag-peak, fig.wide = TRUE, fig.asp = 0.9}
bc_stat[, 'group' := ifelse(bc %in% cell_barcodes, 'cell', 'non-cell')]

nsub_frags = min(15000, nrow(bc_stat))  ## downsample for scatter plot
bc_stat_down = bc_stat[sort(sample(1:nrow(bc_stat), nsub_frags)), ]
g <- ggplot(data = bc_stat_down, 
            aes(x = total_frags, y = frac_peak, col = group)) + 
  geom_point(size = 0.5) + scale_x_continuous(trans='log10') + theme_bw() +
  theme(legend.position = 'none', 
        legend.title=element_blank(),
        axis.text = element_text(size = 15, family = "Helvetica"),
        axis.title = element_text(size = 18, family = "Helvetica")) +
  xlab('Total #Unique Fragments') + ylab('Fraction in Peak')

text1 <- grobTree(textGrob("Cell", x=0.8,  y=0.93, hjust=0,
                           gp=gpar(col='#D95F02', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))
text2 <- grobTree(textGrob("Non-cell", x=0.8,  y=0.83, hjust=0,
                           gp=gpar(col='#7570B3', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))

g + annotation_custom(text1) + annotation_custom(text2) + scale_color_manual(values = c('#D95F02', '#7570B3'))
```

## Distribution of the insert size

A density plot representing the distribution of the insert sizes (fragment size excluding adapters).
```{r insert-size-dist, fig.wide = TRUE, fig.asp = 0.9}
  frags[, 'isize' := V3 - V2]
  if (nrow(frags) >= 100000) {
    frags = frags[sort(sample(1:nrow(frags), 100000)), ]
  }
  
  ggplot(data = frags[isize < 800], aes(x = isize)) +
    geom_density(fill = '#1B9E77') + xlab('Insert Size (bp)') + ylab('Density') + theme_bw() +
    scale_fill_brewer(palette="Set1") +
    theme(legend.title=element_blank(), 
          legend.background = NULL, 
          axis.text = element_text(size = 15, family = "Helvetica"), 
          axis.title = element_text(size = 18, family = "Helvetica"))
```

## Density plot of the total number of unique fragments

A density plot of the total number of unique fragments, comparing between cells and non-cells.
```{r frag-density, fig.wide = TRUE, fig.asp = 0.9}
bc_stat[, 'group' := ifelse(bc %in% cell_barcodes, 'cell', 'non-cell')]

p <- ggplot(data = bc_stat, aes(x = total_frags, fill = group)) + 
  geom_density() + scale_x_continuous(trans = 'log10') + theme_bw() +
  theme(legend.position='none', legend.title=element_blank(),
        axis.title = element_text(size = 18, family = "Helvetica"),
        axis.text = element_text(size = 15, family = "Helvetica")) + 
  xlab('Total #Unique Fragments') + ylab('Density') 

text1 <- grobTree(textGrob("Cell", x=0.8,  y=0.93, hjust=0,
                           gp=gpar(col='#D95F02', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))
text2 <- grobTree(textGrob("Non-cell", x=0.8,  y=0.83, hjust=0,
                           gp=gpar(col='#7570B3', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))

p + annotation_custom(text1) + annotation_custom(text2) +
  scale_fill_manual(values = c('#D95F02', '#7570B3'))
```

## Box plots of overlap with annotated regions

Box plots which show the proportion of fragments in called cells that overlap with various annotated regions.
```{r anno-overlap-boxes, fig.wide = TRUE, fig.asp = 0.9}
qc_sele_df = data.table(frac = c(qc_sele$frac_peak, qc_sele$frac_tss, qc_sele$frac_promoter, qc_sele$frac_enh, qc_sele$frac_mito), 'type' = rep(c('Peaks', 'TSS', 'Promoter', 'Enhancer', 'Mito'), each = nrow(qc_sele)))
  
qc_sele_df$type = factor(qc_sele_df$type, levels = c('Peaks', 'TSS', 'Promoter', 'Enhancer', 'Mito'))

ggplot(data = qc_sele_df, aes(y = frac, x = type, fill = type)) + ylab('Fraction') + theme_bw() +
  geom_boxplot(outlier.size = 0.01, show.legend = FALSE) + 
  scale_fill_brewer(palette="Set1") +
  theme(legend.position = 'none', 
        axis.text = element_text(size = 18, family = "Helvetica"), 
        axis.title.x = element_blank(), 
        axis.title.y = element_text(size = 18, family = "Helvetica")) + xlab('') 
```

## Percentage of fragments overlapping with annotated regions

```{r overall-overlap, fig.wide = TRUE, fig.asp = 0.9}
frac_peak = sum(qc_sele$total_frags * qc_sele$frac_peak)/sum(qc_sele$total_frags)
frac_mito = sum(qc_sele$total_frags * qc_sele$frac_mito)/sum(qc_sele$total_frags)
frac_promoter = sum(qc_sele$total_frags * qc_sele$frac_promoter)/sum(qc_sele$total_frags)
frac_enh = sum(qc_sele$total_frags * qc_sele$frac_enhancer)/sum(qc_sele$total_frags)
frac_tss = sum(qc_sele$total_frags * qc_sele$frac_tss)/sum(qc_sele$total_frags)

fracs = data.frame(c(frac_peak,  frac_promoter, frac_enh, frac_tss, frac_mito))
row.names(fracs) = c('Fraction in peaks', 
                     'Fraction in promoters', 
                     'Fraction in Enhancers(ENCODE)', 
                     'Fraction in TSS',
                     'Fraction in mitochondrial genome')
colnames(fracs) = 'pr'
fracs$pr = round(fracs$pr, 3)
fracs$pr = paste0(100*fracs$pr, '%')

kable(fracs, row.names = T, col.names = NULL) %>%
  kable_styling(full_width = F, position = 'left', font_size = 15)
```

